### PR TITLE
fix(#1785): retain focused node on filter toggle

### LINKED ISSUE #1785: Focus Incorrect After Filter Toggle
### Description

After toggling a filter the current line is unchanged, which can change selected file.

### Neovim version

```text
NVIM v0.8.1
Build type: Release
LuaJIT 2.1.0-beta3
```

### Operating system and version

Linux 6.0.10-arch2-1

### nvim-tree version

829e9f6

### Minimal config

```lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvt-min/site]]
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
  require("packer").startup {
    {
      "wbthomason/packer.nvim",
      "nvim-tree/nvim-tree.lua",
      "nvim-tree/nvim-web-devicons",
      -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. "/plugin/packer_compiled.lua",
      display = { non_interactive = true },
    },
  }
end
if vim.fn.isdirectory(install_path) == 0 then
  print "Installing nvim-tree and dependencies."
  vim.fn.system { "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path }
end
load_plugins()
require("packer").sync()
vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]]
vim.opt.termguicolors = true
vim.opt.cursorline = true

-- MODIFY NVIM-TREE SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
_G.setup = function()
  require("nvim-tree").setup {}
end
```

### Steps to reproduce

open nvim-tree.lua
open doc
move cursor to `CONTRIBUTING.md`
`I`

### Expected behavior

cursor at `CONTRIBUTING.md`

### Actual behavior

cursor at `.stylua.toml`

### PR DESCRIPTION
**Pull request by Akmadan23**
fixes #1785 

I was annoyed by the fact that after clearing the live filter the cursor does not follow the focused node, so I made a simple change that addresses the issue. Maybe it could be enabled or disabled with an option, something like `live_filter.focus_node_on_clear`. Let me know if I should add it or if it's reasonable to leave it as is.

### CODE REVIEW

**Comment by gegoune** [comment]
Thanks for pr. This has been worked on by @alex-courtis in https://github.com/nvim-tree/nvim-tree.lua/pull/1872 but no satisfying solution was found at the time. Would you mind having a look at linked pr to check if concerns raised there are not affecting your solution here?

**Comment by Akmadan23** [comment]
> If file is not opened cursor will jump to first line of tree and remain there when switching filter on and off. Seem like it just doesn't have an entry to jump to.

This is not happening to me, maybe because I'm using `utils.focus_file` passing the full path of the node, so that if it's visible, it simply focuses that. Or maybe it's because I'm only applying it to the live filter instead of the various filters that Alex applied it to, but I can assure it works fine. Or maybe I completely misunderstood the issue here

**Comment by Akmadan23** [comment]
I might have found a solution, I'll paste it here in the comment so that we can discuss it and then I will commit to the PR if it will be approved.

```lua
local function reload()
  local node = lib.get_node_at_cursor()
  reloaders.reload_explorer()

  while node do
    local visible_nodes = require("nvim-tree.core").get_explorer().nodes
    local found_node, _ = utils.find_node(visible_nodes, function(n)
      return n.absolute_path == node.absolute_path
    end)

    if found_node or node.parent == nil then
      utils.focus_file(node.absolute_path)
      break
    end

    node = node.parent
  end
end
```

This `reload` function is applied exactly like @alex-courtis did in #1872, and according to my tests it fixes all the issues presented above. Let me know if it actually makes sense or if I'm missing something important here.

By the way, all this tests are completely useless applied the live filter, because when it gets cleared nodes cannot disappear but only reappear, so if a node has focus it will keep it with no edge cases. That's why at first I thought it was immune to that problem.

**Comment by alex-courtis** [comment]
> I might have found a solution, I'll paste it here in the comment so that we can discuss it and then I will commit to the PR if it will be approved.

I'm excited!

Please push or create a new branch and I'll have a play with it.

File: lua/nvim-tree/actions/tree-modifiers/toggles.lua [hunk 1]
```diff
@@ -1,31 +1,52 @@
+local lib = require "nvim-tree.lib"
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/tree-modifiers/toggles.lua:1]:
```suggestion
local core = require "nvim-tree.core"
local lib = require "nvim-tree.lib"
```

File: lua/nvim-tree/actions/tree-modifiers/toggles.lua [hunk 2]
```diff
@@ -1,31 +1,52 @@
...
+local utils = require "nvim-tree.utils"
 local filters = require "nvim-tree.explorer.filters"
 local reloaders = require "nvim-tree.actions.reloaders.reloaders"
 
 local M = {}
 
+local function reload()
+  local node = lib.get_node_at_cursor()
+  reloaders.reload_explorer()
+  local visible_nodes = require("nvim-tree.core").get_explorer().nodes
```
Comment (alex-courtis) [inline lua/nvim-tree/actions/tree-modifiers/toggles.lua:11]: Please null test `get_explorer()`

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Test cases:
- [x] live filter
- [x] B
- [x] C
- [x] H
- [x] I
- [x] U
- [x] focus parent - child not selected
- [x] focus closed parent
- [x] child node selected
- [x] group_empty

**Comment by alex-courtis** [comment]
This feels good to me, and it's how I would build the tree from scratch today.

1 "focus root" is different to current behaviour
3 "do nothing" is current behaviour but it is odd

I'm happy either way. Your call @gegoune !

**Comment by Akmadan23** [comment]
About the requested changes, is it better to add another commit or do you prefer a force push with a single commit (in my opinion much clearer)?

**Comment by gegoune** [comment]
Please push individual commits, we can squash on merge.

@alex-courtis either way, really, if it works properly and consistently.

**Comment by gegoune** [comment]
I gave it a very quick test - doesn't seem to work very well for me. After clearing filters it is parent directory that gets focus quite often. Clearing live filter lives cursor at the position where live filtering input was.

**Comment by Akmadan23** [comment]
> After clearing filters it is parent directory that gets focus quite often.

I'm not sure what you mean here...

> Clearing live filter lives cursor at the position where live filtering input was.

Yeah, I didn't consider that scenario, the easiest solution is to focus root, but we could also save the last focused node in a local variable and then fall back to that if no node is found under the cursor on clear.

Moreover, since also `collapse_all` has this issue, I think would be handy to move the `reload` function to the `utils` module, something like `utils.focus_node_or_parent(node)`, so that it can be easily reused without rewriting it. Wouldn't it be nice?

**Comment by alex-courtis** [comment]
> > 
> 
> Yeah, I didn't consider that scenario, the easiest solution is to focus root, but we could also save the last focused node in a local variable and then fall back to that if no node is found under the cursor on clear.

That would feel nice, quite similar to current behaviour, but actually correct in both cases.

**Comment by alex-courtis** [comment]
> Moreover, since also `collapse_all` has this issue, I think would be handy to move the `reload` function to the `utils` module, something like `utils.focus_node_or_parent(node)`, so that it can be easily reused without rewriting it. Wouldn't it be nice?

That sounds good. Can we please do that refactor in a separate PR after this?

**Comment by Akmadan23** [comment]
> Can we please do that refactor in a separate PR after this?

Sure, for now I'm pushing the fix for the live filter issue which is actually really simple.

**Comment by alex-courtis** [comment]
This is fantastic!

I'll dogfood this one for a week and get a feel for it.

File: lua/nvim-tree/live-filter.lua [hunk 3]
```diff
@@ -2,6 +2,9 @@ local view = require "nvim-tree.view"
 local utils = require "nvim-tree.utils"
 local Iterator = require "nvim-tree.iterators.node-iterator"
 
+-- Keeps memory of focused node when entering filter prompt
+local last_node
```
Comment (alex-courtis) [inline lua/nvim-tree/live-filter.lua:6]: This will be a future problem when we finally sort out multiple trees / tabs.

Can you please store this at `view.View`?

**Review by alex-courtis — CHANGES_REQUESTED** [review]
This is looking really good... we are close.

**Comment by alex-courtis** [comment]
Dogfooding for a week.

**Review by alex-courtis — APPROVED** [review]
Many thanks for your contribution.

**Comment by Akmadan23** [comment]
It was a pleasure.
